### ISSUE TITLE
Proposal: allow generic methods to specify T? without constraining to class or struct

### ISSUE THREAD

**Issue by roji** [dotnet/csharplang#2194]
Currently, specifying `T?` in a generic method signature requires it to be constrained to either a struct or class. The proposal is to allow the following without any constraints:

```c#
T? Foo<T>() {...}
```

For value types, the method would return `Nullable<T>` as usual, and for reference types it would return a nullable reference for the purposes of compiler analysis.

As an example, when looking at a better (and more null-friendly) API for fetching database column values on DbDataReader, I am currently blocked by the inability to define a generic method that returns `T?`, and that can work on both value and reference types. The only alternative would be to define two methods - one constrained on `class` and the other on `struct` - but since overloading on generic constraints isn't supported they must have different names. This would force the user to deal with the value/reference type distinction - figuring out which method to call - and seems like a non-starter for the kind of API we're looking for.

More generally, it seems that C# 8 nullability introduces more difficulties when intersecting with the value/reference type divide - developers are unfortunately forced to think about this distinction more than before nullability existed. It seems like it's important to smooth that out (see #1865 in general for diminishing the feature gap). As an example, C# already allows one to use (nullable) references and `Nullable<T>` interchangeably in several contexts: the null coalescing and null-conditional operators accept both. It would be good to go further in this direction.

This issue may be a dup of #2146, although that issue seems a bit unclear. See also https://github.com/dotnet/roslyn/issues/30953 which is somewhat related, although that issue is about returning default, whereas here the idea is to allow returning null for both reference and value types.

/cc @divega @ajcvickers

**Comment by johnkellyoxford** [comment]
This isn't currently possible in C# 8?!

**Comment by roji** [comment]
> This isn't currently possible in C# 8?!

Not yet :)

**Comment by HaloFour** [comment]
> For value types, the method would return `Nullable<T>` as usual, and for reference types it would return a nullable reference for the purposes of compiler analysis.

How do you propose that actually work?  A method can't selectively change its return value.  A generic method has the same signature and body regardless of the generic type arguments, it can't return `T` sometimes and `Nullable<T>` other times.

**Comment by DanFTRX** [comment]
@HaloFour It could generate 2 methods at compile time, one with the struct constraint and one with the class constraint

**Comment by DanFTRX** [comment]
@HaloFour Could it possibly generate 2 methods at compile time, one with a Nullable<T> constrained to struct and one with a nullable reference?  I know it would be valid language code to have two different return types with same function name due to ambiguity, but is that a runtime restriction or a compile restriction? (I know its just a compile restriction for Java as I've done it in the past, not sure about .NET)

**Comment by roji** [comment]
You're right of course - this isn't just a compiler-level feature, it would require deeper changes in other places. For example, the combination of a `[Nullable]` attribute along with the absence of a struct/class constraint could be interpreted (by the JIT, by the compiler of consuming code...) as meaning `Nullable<T>` when T is a struct, or T? when it is a class. Unfortunately that would be a breaking change as older compilers will only see a T return value. I'm hoping someone more knowledgeable can take this further.

I do realize that on the implementation level this is a pretty big/complicated ask, I'm only expressing the need at the language level. I'm mostly concerned that nullability is making things significantly more complicated than before.

PS If overloading on generic constraints were allowed (#2013) this would be much less of a problem, but that doesn't seem to be adopted in any way either.

**Comment by roji** [comment]
@DanFTRX the problem with generating two methods is that they would collide - same name with the same parameters...

**Comment by HaloFour** [comment]
I think your only option would be for the compiler to emit two different versions of the method.

The CLR does allow overloads based on different returning types, or even just on the application of `modopt`/`modreq` modifiers on parameters.  My concern would be how they would work with downlevel compilers or other languages.

**Comment by Joe4evr** [comment]
I think that this requires a more configurable approach than what is proposed. Consider that there are also various open generic methods that *don't* return a `Nullable<T>` for value types, e.g. the `XxxOrDefault()` methods in LINQ. In order to properly integrate with the Nullability Analysis, a new recognized attribute should be able to be applied to open generic methods to specify how the return value has to be interpreted. I'm thinking something like this:
```cs
[return: OpenNullable(ValueTypeHandling.DefaultValue)] //compiler emits one method and in consuming code ignores the nullability flow if invoked with 'T' as a value type
public T? FirstOrDefault<T>(this IEnumerable<T> sequence);

[return: OpenNullable(ValueTypeHandling.Nullable)] //compiler emits two methods as suggested by HaloFour
public T? ReadValue<T>();
```

**Comment by roji** [comment]
@Joe4evr what you're thinking about seems to be covered by https://github.com/dotnet/roslyn/issues/30953?

Methods like `xxxOrDefault()`, which don't return a `Nullable<T>` for value types, shouldn't have a return value of `T?` - this has been a synonym for `Nullable<T>` for a long time. The current plan seems to be to handle your case with a `[MaybeNull]` attribute, which would mean that T is nullable only if it's a reference type.

**Comment by omariom** [comment]
If it is not a virtual method, this approach could help
https://twitter.com/xoofx/status/1087770199854665729?s=20

**Comment by JeroenBos** [comment]
@roji: the name and parameters would collide yes, but what if type parameter constraint also distinguished? That would neatly separate these two cases. The method with `class` constraint would return `T`, and with `struct` would return `T`?

**Comment by Logerfo** [comment]
@JeroenBos that would be a breaking change, since the constraints are not considered as part of the method signature today. Changing this would make the new language spec only available for a new runtime version.

**Comment by JeroenBos** [comment]
@Logerfo  Agreed. I was thinking of [#2196](https://github.com/dotnet/csharplang/issues/2196), where the runtime requirement it is also mentioned. I think my  point is that _if_ that proposal were to be implemented, this proposal could be implemented over that path.

**Comment by HaloFour** [comment]
The return type is a part of the method signature per the CLR, so there'd be no collision.

**Comment by DanFTRX** [comment]
@roji I mean I just said that ._.

**Comment by roji** [comment]
@DanFTRX my bad, sorry - I wasn't aware this was possible at the CLR level when I wrote the comment above. So this seems like it could actually be a compiler-only feature.

The question still remains of what happens when an older compiler encounters this situation, though...

**Comment by Thaina** [comment]
Maybe we could have `Maybe<T>` that work completely the same as `Nullable<T>` but can accept class and con convert to `Nullable<T>` of struct

And then having `T?` without constrain become `Maybe<T>` instead

**Comment by roji** [comment]
@Thaina the problem with that is that it introduces yet another representation of something nullable, in addition to one already being introduced in C# 8... It's also not (directly) compatible with any of the other representations...

**Comment by sharwell** [comment]
I would prefer `T?` for unconstrained generic type parameter `T` be interpreted as `T` for non-nullable value types and `T?` otherwise. This avoids changes in the underlying return type between instantiations of the generic code and can be implemented using CLR 2+ rules.

**Comment by roji** [comment]
`T?` for value types already means something very clear in C# - `Nullable<T>` - and it has meant that for many years, I don't see how it can suddenly mean `T` (even in a new context) that would extremely confusing and inconsistent..

**Comment by roji** [comment]
Also, read the above on possibilities sorry implementing this without runtime changes (i.e. two overloaded methods with different result types). This is a new language feature which doesn't really have to conform to older rules.

**Comment by YairHalberstadt** [comment]
@roji

Originally I agreed with you, that `T?` should mean Nullable<T> on an unconstrained T.

However in my experience using NRTs I've noticed that the main times I want to express T?, are when I want to write default (T), but can't because T may be a value type. For example in FirstOrDefault.

For that reason I'm leaning towards @sharwell s opinion on this.

I would also accept that it's just too confusing, and shouldn't be done at all.

**Comment by orthoxerox** [comment]
@YairHalberstadt I can understand that, but I, on the other hand, have mostly encountered use cases where I wanted T? to be T?, that is, either `default(Nullable<T>)` or `null`.

**Comment by YohDeadfall** [comment]
Just my two cents. There is the same feature in C++ where [`optional<T>`](https://en.cppreference.com/w/cpp/utility/optional) exists to represent a nullable value. Yep, it doesn't support reference types yet, but things may change in the future ([yes, we need it](https://medium.com/@barryrevzin/optional-t-yes-we-need-it-74bf5958b81)). How is it related to this discussion? Just express the code above in C++ and think how absurdly it will look like (:

So it's better to constrain `T` to be nullable rather than go in opposite way. For example, the `BaseGenericClass<T>` doesn't limit us, but `NestedGenericClass<T>` explicitly says that `T` may have no value. Otherwise, the idea of a generic parameter will be changed, it won't mean a placeholder or anything.

Another downside of having to mark an unconstrained generic parameter with `?` is having it mostly everywhere (`IEnumrable<T>` and other collection types, `Action<T?> and other delegates, and so on). Simply it will pollute code and make it uglier.

**Comment by roji** [comment]
@YairHalberstadt and others, there really seem to be valid use cases for expressing two things:

1. A default value (i.e. null for reference, 0 for int, e.g. the return value of `FirstOrDefault()`), this is what's covered by `[MaybeNull]` in https://github.com/dotnet/roslyn/issues/30953
2. A nullable value (`int?` and `string?` returned by the same method). This is covered by this issue.

If we redefine `T?` to start meaning "default" in an unconstrained generic context, we lose the ability to ever have the 2nd option. More importantly, there already is a reasonable-looking proposal for the 1st method - simply adding `[MaybeNull]` on the type parameter - so what's the advantage of a solution that makes the meaning of `T?` ambiguous and context-dependent, and also precludes ever adding the 2nd option?

**Comment by YohDeadfall** [comment]
I don't find the idea about the `[MaybeNull]` attribute good, because every time you need to limit a parameter, you have to write a long sequence of letters. Things can be made much simpler if an existing language feature gets reused. How are you about `T!` in a generic type definition? This language construction already means that the value is not `null`, and the same meaning can be applied to types, so:

* `T` means the same thing as before, it's an unconstrained type which could be either nullable or non-nullable.
* `T!` means that the type is non-nullable.
* `T?` means that the type is nullable.

**Comment by roji** [comment]
@YohDeadfall your proposal suffers from the same kind of issue as the previous one:

* When the C# 8 nullability feature is turned on, `T` means "non-nullable" in all contexts. You would make it means something else in the unconstrained generic context.
* `T!` is currently not a type: you cannot have a variable of type `T!`, and `!` is an operator for suppressing compiler nullability warnings.

I agree that having to write out a full attribute isn't ideal, even if it would only have to be specified once on the definition (and not every time T is used). But any proposal that redefines existing syntax to mean something else just in the unconstrained generic context - just to make the attribute shorter - seems like a problematic direction to me.

**Comment by YohDeadfall** [comment]
> But any proposal that redefines existing syntax to mean something else just in the unconstrained generic context - just to make the attribute shorter - seems like a problematic direction to me.

The current trend is to make construction simpler. Remember patter matching, for example, or the null coalescing assignment operator.

I think that having `?` as a may-be-nullable constraint is a good idea. It already exists and it's the `class?` constraint, but it's for reference types only and means nullable references are valid too. The following example demonstrates its behavior:

```csharp
public static void DoWithNullable<T>(T value) where T : class?
{
}
    
public static void DoWintNonNullable<T>(T value) where T : class
{
}
    
public static void Main(string[] args)
{
    DoWithNullable<string?>(null); // OK
    DoWithNullable<string>("value"); // OK
       
    DoWintNonNullable<string?>(null); // Warning CS8634: Nullability of type argument 'string?' doesn't match 'class' constraint.
    DoWintNonNullable<string>("value"); // OK
}
```

The problem here is that there is no way to express the same thing for a struct and for any type (either reference or value). So having `struct?` and `any?` constraints could solve the issue, but at some cost. This is what @sharwell proposed.

**Comment by YohDeadfall** [comment]
Just for you information, `class?` is a syntax sugar which adds an attribute:

```csharp
public static void DoWithNullable<[Nullable(2)] T>([Nullable(1)] T value) where T : class
{
}

public static void DoWintNonNullable<[Nullable(1)] T>([Nullable(1)] T value) where T : class
{
}
```

@jcouv Could you explain the meaning of magic values which `NullableAttribute` accepts?

**Comment by yaakov-h** [comment]
```csharp
public static void DoWintNonNullable<[Nullable(1)] T>([Nullable(1)] T value) where T : class
{
}
```

This is incorrect. There is [no `[Nullable]` attribute emitted](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEDMACFAmDBhAWACgBvYjCrTFARgDYsAWDAEQHsB1ASwDsYA5Nj34BXADZiIwMXAA8AFQB8ACnkYAbhDEi4ASgwB3ABaI4GNSCzZylMkUoYAvsUdA=) in this case.

**Comment by YohDeadfall** [comment]
To see it you should do a little trick, [add `#nullable enable`](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4GIB2BXANliwtwAEcaeBAsAFAACAzIdQEyEDCVA3lYdw/dQIwA2BgBZCAEQD2AdQCWaGADlJaRdlz44AHgAqAPgAUOwgDcIWDHACUhAO4ALREWMgGjLj06UehAL5VfIA===) (:

**Comment by yaakov-h** [comment]
ah right, without that it gets emitted as oblivious...

and therein lies the answer!

> Each type reference is accompanied by a NullableAttribute with an array of bytes, where 0 is Oblivious, 1 is NotAnnotated and 2 is Annotated. All value types are marked with flag 0 (oblivious).

https://github.com/dotnet/roslyn/blob/910a3331f730a435c3ced92888568924a789771c/docs/features/nullable-reference-types.md

**Comment by YohDeadfall** [comment]
@yaakov-h Thanks for answering the question to Julien! Out of curiosity, why is `byte` used instead of an `enum`?

**Comment by yaakov-h** [comment]
No idea. At a guess, to reduce the amount of code that needs to be embedded in the target assembly?

**Comment by roji** [comment]
See https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-05-15.md#unconstrained-t

**Comment by roji** [comment]
@jcouv in [this section or the recent nullability LDM](https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-05-15.md#unconstrained-t), it's stated that "There is therefore no real need for the ability to say T? on an unconstrained T`, since `[AllowNull]` and `[MaybeNull]` can be specified together.

However, the point of unconstrained T? (at least in this issue) is to have `Nullable<T>` when T is a value type, without having to constrain the generic method to value types only. Is this concern also being addressed in some way?

**Comment by jcouv** [comment]
@roji There are two views of what `T?` might mean. One view is that it means `Nullable<T>` when `T` is a struct, and the other is that it doesn't change the type (it's still a runtime `T`) but provides an annotation for reference types.

So far in the nullable feature, annotations do not change emitted/runtime types, so we mostly focused on the latter interpretation. 
In the former interpretation, it is not even clear what would be emitted in IL: is it a `T` or a `Nullable<T>`? 
It might be possible to do this with some runtime support, but it's not clear yet that the ability to express `T or Nullable<T>` as a type is that useful or desirable given the complexity.

**Comment by roji** [comment]
@jcouv thanks for the clarification.

Yeah, this issue is about `T?` meaning `Nullable<T>` for value types (the former interpretation), and above are some ideas on how that could happen (e.g. compiler emitting two methods - one for reference types and one for value types - overloaded on the return type). It's indeed complex, but it does seem highly desirable to have a way for a generic method to specify a nullable "something", without the user having to care whether it's a value or reference type.

The "latter interpretation", where `T?` is `T` in runtime seems problematic in that `T?` means `Nullable<T>` in every other context when `T` is a value type. It would be odd if `T?` meant something different in an unconstrained generic context and in a struct-constrained context. In addition, with the new `[MaybeNull]` and `[AllowNull]` it's not clear why this would even be necessary as the attributes suffice.

**Comment by KillyMXI** [comment]
I want to bring my perspective on nullable return types, especially `out` parameters. I believe this issue is the best fitting one to add to, but let me know if it is better to open a new issue.

When writing generic functional code, some language legacy forces us to write duplicate code. Most notorious example is when we have to implement the same code for `IEnumerable<T>` and `T[]`. Iirc, authors of extension libraries such as LanguageExt can bring up more examples, but that's beyond the point of this comment so I won't look it up now.

From what I see, the different handling of `Nullable<T>` and nullable reference types will surpass the pain of duplicating code existed before.

And what I find rather ironic, the new language feature doesn't work well with the established `Try` pattern.

----

 **Example 1:**
 
 (Taken from [this StackOverflow question](https://stackoverflow.com/questions/53759235/out-parameters-with-nullable-reference-types/).)
 
 Let's say we write an implementation of the Option/Maybe monad and want to provide a convenience method to make it integrate better into existing code by implementing a common .Net pattern...
 
 ```csharp
public abstract Option<T> {} // exact implementation skipped as irrelevant
public class None<T> : Option<T>
public class Some<T> : Option<T> 

public static class OptionExtensions
{
    public static bool TryGetValue<T>(this Option<T> option, out T value)
    {
        if (option is Some<T> some)
        {
            value = some.Value;
            return true;
        }

        value = default;
        return false;
    }
}
```

Once we add [NotNullWhenAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullwhenattribute?view=netstandard-2.1) to `out T value`, we have to make the choice - do we need T to be a reference or value type.

But we need both of them. We don't want to narrow it down!

Thanks to @omariom for bringing up the workaround. Since the issue #1628 was fixed, we now can have multiple extensions that differ only in `where` constraints within the same class. But we still have to have duplicate extension methods for classes and structs.

For this particular example it looks acceptable. Just two copies...

First, we introduced non-trivial code duplication. I suspect there will often be nontrivial code, not shareable between these extension classes, or otherwise there must be a better workaround...

Second, what if we have more than one nullable type in the return values? More examples for that:


**Example 2:**

Not going too far from our monads. My own issue after which I started to look around and found the above-mentioned SO question - I wanted to do the same thing with the Either monad.

```csharp
public abstract class Either<L, R>
{
    // most of the implementation is skipped as irrelevant
    public abstract T Match<T>(Func<L, T> f, Func<R, T> g);

    private class Left : Either<L, R>{ /* skipped */ }
    private class Right : Either<L, R>{ /* skipped */ }
}

public static class EitherExtensions
{
    public static bool TryGetValue<L, R>(
        this Either<L, R> either,
        [NotNullWhen(false)] out L? left,
        [NotNullWhen(true)] out R? right
        )
        where L : class
        where R : class
    {
        L? tempLeft = null;
        R? tempRight = null;
        bool result = either.Match(
            l => { tempLeft = l; return false; },
            r => { tempRight = r; return true; }
            );
        left = tempLeft;
        right = tempRight;
        return result;
    }
    
    /* Three more of the same are skipped for brevity.
       Thankfully I can do this in an example code... */
}
```

I have two independent types - `L` and `R`. Even if I expect `L` to be almost always a `string`, I don't want to have a pressure to choose between code amount and universality. So I had to make **four** copies of the extension method - 68 lines of code instead of 17.


**Example 3:**

I have a bunch of extension methods to work with nullables as poor man's Option...

No `out` parameters here, but the issue is the same - compiler has to know which kind of nullables there are, without any code differences.

```csharp
public static R? MapNullable<T, R>(this T? item, Func<T, R?> fn)
    where T : struct
    where R : struct
    /* !item.HasValue ? null : fn(item.Value) */
    => item == null ? null : fn(item.Value);

public static R MapNullable<T, R>(this T? item, Func<T?, R> fn)
    where T : struct
    => fn(item);

public static R? MapNullable<T, R>(this T? item, Func<T?, R?> fn)
    where T : struct
    where R : struct
    => fn(item);

// ... few more MapNullable<> variations skipped
```

This time, contents of the functions is trivial. But the amount of them makes it hard to maintain them if I want to have them universal across all class and struct combinations. Some of them have to be copied twice, some - four times... Some - precisely, some - not quite. And it all adds up.

Granted, I'm not heavily dependent on these methods and can get rid of them, but the same issue will surface in one shape or another for different people.


----

I'm waiting to run into a problem when there will be a need for a function with a combination of multiple different input types, return/out types, collection kinds to see how big of a combinatorial explosion we can realistically get.

It seems like commenters in the issue #2309 also had a short discussion on what can be done about this, but also without any conclusion.

This is too much of PITA for programmers to handle this manually. If there is a single point to remove the frustration for a lot of people - it must be done.

This issue need to have higher priority and worked on more actively. I hope to see more in-depth discussion about implementation possibilities and roadblocks.

----

For input parameters the challenge seems to be to reconcile some existing API's. We are expected to operate with `Nullable<T>` using it's `HasValue` and `Value` properties, while there is no such thing for bare reference types.

I'm not a fan of an idea to introduce a `ReferenceNullable<T>` wrapper for classes - it will just complicate things further. I think something has to be done about `Nullable<T>` to bring it closer to the new nullable behavior and let that work when it is known only as `T?`. We don't have to rely on `HasValue` already. Teaching the compiler to replace `Nullable<T>` with the underlying value when appropriate must be more challenging task though...

The limitation for output types seems to be just a collateral damage from that. Correct me if I'm missing something.

If the limitation can be removed just for return types at least - it will be really helpful already. I suspect there are also challenges in the inner workings to make that possible. I would like to hear a competent feedback on that. Is it common for C# compiler to emit multiple copies of a function for different underlying types?

**Comment by roji** [comment]
This seems like it's covered by https://github.com/dotnet/roslyn/issues/29146, closing for now.

**Comment by roji** [comment]
Read that too quickly, https://github.com/dotnet/roslyn/issues/29146 is about expressing the default of T for unconstrained generic types. This proposal is about being able to express a single generic method where T? (or some other syntax) express Nullable<T> for value types, and NRTs for reference types. In other words, T? would refer to two different *runtime* types depending on whether T is a value or reference type (int? for int, string for string).

Reopening even though there seems to be little chance this will be implemented (https://github.com/dotnet/csharplang/issues/2194#issuecomment-498095926).
